src/platform/i386/lapic.c

```diff
@@ -357,10 +404,99 @@ lapic_timer_init(void)
 
 	/* Set the divisor */
 	lapic_write_reg(LAPIC_DIV_CONF_REG, LAPIC_DIV_BY_1);
+
+	lapic_ack();
+}
+
+static int
+lapic_ipi_send(u32_t dest, u32_t vect_flags)
+{
+	lapic_write_reg(LAPIC_ICR + 0x10, dest << 24);
+	lapic_read_reg(LAPIC_ICR + 0x10);
+
+	lapic_write_reg(LAPIC_ICR, vect_flags);
+	lapic_read_reg(LAPIC_ICR);
+
+	return 0;
+}
+
+/* HACK: assume that the HZ of the processor is equivalent to that on the computer used for compilation. */
+static void
+delay_us(u32_t us)
+{
+	unsigned long long hz = CPU_GHZ, hz_per_us = hz * 1000;
+	unsigned long long end;
+	volatile unsigned long long tsc;
+	rdtscll(tsc);
+	end = tsc + (hz_per_us * us);
+	while (1) {
+		rdtscll(tsc);
+		if (tsc >= end) return;
+		asm("pause");
+	}
+}
+
+/* The SMP boot patchcode from loader.S */
+extern char smppatchstart, smppatchend, smpstack, stack;
+
+void
+smp_bootall(volatile int *cores_ready)
+{
+	int i;
+	u32_t ret;
+	char **stackpatch;
+
+	/*
+	 * Set up the processor boot-up code .  Use SMC to create the
+	 * smp loader code with the stack address inlined into it at
+	 * an address that real-mode 16-bit code can execute
+	 */
+	memcpy((char *)chal_pa2va(SMP_BOOT_PATCH_ADDR), &smppatchstart, &smppatchend - &smppatchstart);
+	stackpatch  = (char **)chal_pa2va(SMP_BOOT_PATCH_ADDR + (&smpstack - &smppatchstart));
+
+	for (i = 1 ; i < ncpus ; i++) {
```

Comment:
Should be `for (i = 1; i < ncpus; i++)`. No space before `;`.

Reply:
my fault ;-)

Reply:
dito.